### DISCUSSION TITLE
Implementing TreeView in an extension

### DISCUSSION THREAD

**Discussion by NoahTheDuke** [neoclide/coc.nvim#4880] [Q&A]
I'm trying to implement TreeView in [coc-clojure](https://github.com/NoahTheDuke/coc-clojure), and the documentation + types isn't enough for me to figure it out. There aren't any examples in the tests, just a lot of unit tests. Has anyone implemented TreeView in an extension so I can see an example?

**Comment by fannheyward** [comment]
coc-clangd uses tree virw to display AST, hope this can help you https://github.com/clangd/coc-clangd/blob/master/src/ast.ts

**Reply by NoahTheDuke** [reply]
Thank you, that's helpful. I've implemented a working TreeView for coc-clojure with that as an example, but I have a few more questions.

1) What is the list of commands for the returned `item` in `getTreeItem`? I have borrowed clang-d's "Jump to", which works for items that have ranges, but if they don't, clicking on an item throws an error which is shown in the status line, saying `Error on notification doKeymap Error: Failed to resolve command from TreeItem.` How do I avoid this? Is there a `no-op` command I can send to do nothing?

2) `setRoot` is called when initializing the Tree View. Is it (and the associated event emitter) necessary? I'm not sure what it's providing (other than being able to say `return [this.root];` in `getChildren`).

3) Would you be interested in a small tutorial/set of docs for writing one's own TreeView? The existing documentation/examples are pretty threadbare, and I'm willing to help.

**Reply by hexh250786313** [reply]
@NoahTheDuke 

1. `item` is a instance of `TreeItem` which define every nodes of treeview. Its command means the action you trigger in it and you can use `commandsMap` to define your custom commands. For example, https://github.com/hexh250786313/coc-todo-tree/blob/60dd8a86eeb899c9abe6e2b6e4f3ef49067b3555/src/tree.ts#L276 will trigger a `open` command defined in https://github.com/hexh250786313/coc-todo-tree/blob/60dd8a86eeb899c9abe6e2b6e4f3ef49067b3555/src/tree.ts#L111 which actually do nothing. So if you don't do anything for some tree nodes you can handle it like what I done.
2. My treeview  extension [coc-todo-tree](https://github.com/hexh250786313/coc-todo-tree) doesn't have `setRoot`, so I think it is not necessary.

**Reply by NoahTheDuke** [reply]
Thank you, that's very helpul.
